java/google/registry/rde/imports/RdeHostLinkAction.java

```diff
@@ -0,0 +1,160 @@
+import static google.registry.model.registry.Registries.findTldForName;
+import static google.registry.util.PipelineUtils.createJobPath;
+
+import com.google.appengine.tools.mapreduce.Mapper;
+import com.google.common.base.Joiner;
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.common.net.InternetDomainName;
+
+import com.googlecode.objectify.Key;
+import com.googlecode.objectify.VoidWork;
+import google.registry.config.RegistryConfig.Config;
+import google.registry.mapreduce.MapreduceRunner;
+import google.registry.model.domain.DomainResource;
+import google.registry.model.host.HostResource;
+import google.registry.request.Action;
+import google.registry.request.Parameter;
+import google.registry.request.Response;
+import google.registry.util.FormattingLogger;
+import google.registry.xjc.JaxbFragment;
+import google.registry.xjc.rdehost.XjcRdeHost;
+import google.registry.xjc.rdehost.XjcRdeHostElement;
+import org.joda.time.DateTime;
+
+import javax.inject.Inject;
+
+/**
+ * A mapreduce that links hosts from an escrow file to their superordinate domains.
+ *
+ * <p>Specify the escrow file to import with the "path" parameter.
+ */
+@Action(path = "/_dr/task/linkRdeHosts")
+public class RdeHostLinkAction implements Runnable {
+  private static final FormattingLogger logger = FormattingLogger.getLoggerForCallerClass();
+
+  private final MapreduceRunner mrRunner;
+  private final Response response;
+  private final String importBucketName;
+  private final String importFileName;
+  private final Optional<Integer> mapShards;
+
+  @Inject
+  public RdeHostLinkAction(
+      MapreduceRunner mrRunner,
+      Response response,
+      @Config("rdeImportBucket") String importBucketName,
+      @Parameter("path") String importFileName,
+      @Parameter(PARAM_MAP_SHARDS) Optional<Integer> mapShards) {
+    this.mrRunner = mrRunner;
+    this.response = response;
+    this.importBucketName = importBucketName;
+    this.importFileName = importFileName;
+    this.mapShards = mapShards;
+  }
+
+  @Override
+  public void run() {
+    response.sendJavaScriptRedirect(createJobPath(mrRunner
+        .setJobName("Link hosts from escrow file")
+        .setModuleName("backend")
+        .runMapOnly(
+            new RdeHostPostImportMapper(),
+            ImmutableList.of(new RdeHostInput(mapShards, importBucketName, importFileName)))));
+  }
+
+  /**
+   * Return the {@link DomainResource} this host is subordinate to, or null for out of zone hosts.
+   *
+   * @throws {@link IllegalStateException} for hosts without superordinate domains
+   */
+  private static DomainResource lookupSuperordinateDomain(InternetDomainName hostName, DateTime now) {
+    Optional<InternetDomainName> tld = findTldForName(hostName);
+    // out of zone hosts cannot be linked
+    if(!tld.isPresent()) {
+      return null;
```

Comment:
This method's return type should either be Optional or annotated with @Nullable.